apps/qubit/modules/user/actions/clipboardExportAction.class.php

```diff
@@ -131,21 +123,136 @@ public function execute($request)
 
     $this->response->addJavaScript('exportOptions', 'last');
 
-    $this->title = $this->context->i18n->__('Export');
+    $this->title = $this->context->i18n->__('Clipboard export');
+
+    if (!$request->isMethod('post'))
+    {
+      return;
+    }
+
+    $this->response->setHttpHeader('Content-Type', 'application/json; charset=utf-8');
+
+    $this->form->bind($request->getPostParameters());
+
+    if (!$this->form->isValid())
+    {
+      $this->response->setStatusCode(400);
+      $message = $this->context->i18n->__('Invalid export options.');
+
+      return $this->renderText(json_encode(array('error' => $message)));
+    }
+
+    $slugs = $request->getPostParameter('slugs', []);
+
+    if (empty($slugs))
+    {
+      $this->response->setStatusCode(400);
+      $message = $this->context->i18n->__('The clipboard is empty for this entity type.');
+
+      return $this->renderText(json_encode(array('error' => $message)));
+    }
+
+    $this->processForm();
+
+    // Create array of selections to pass to background job where 
+    // Term ID will be key, and Term description is value
+    $levelsOfDescription = array();
+    foreach ($this->levels as $value)
+    {
+      $levelsOfDescription[$value] = $this->choices[$value];
+    }
+
+    $options = array(
+      'params' => array('fromClipboard' => true, 'slugs' => $slugs),
+      'current-level-only' => !$this->descendantsIncluded,
+      'public' => !$this->draftsIncluded,
+      'objectType' => $this->objectType,
+      'levels' => $levelsOfDescription
+    );
+
+    $options['name'] = $this->context->i18n->__('xml' === $this->formatType ? 'XML export' : 'CSV export' );
+
+    if($this->includeDigitalObjects)
+    {
+      $options['name'] = $this->context->i18n->__('%1% and %2%', 
+        array(
+          '%1%' => sfConfig::get('app_ui_label_digitalobject'),
+          '%2%' => $options['name']
+        )
+      );
+      $options['includeDigitalObjects'] = true;
+    }
+
+    // When exporting actors, ensure aliases and relations are also exported
+    if ('actor' === $this->objectType && 'csv' === $this->formatType)
+    {
+      $options['aliases'] = true;
+      $options['relations'] = true;
+    }
 
-    if ($request->isMethod('post'))
+    try
     {
-      $this->form->bind($request->getPostParameters());
+      $jobName = $this->getJobNameString();
 
-      if ($this->form->isValid())
+      // Preview export before attempting, if job supports this
+      $exportPreviewSupported = method_exists($jobName, 'exportOrCheckForResults');
+
+      if ($exportPreviewSupported && !call_user_func(array($jobName, 'exportOrCheckForResults'), $options))
```

Comment:
Not really part of this work, but I'll check it out.

Reply:
I agree that it's not the easiest to understand but I think it's okay ...

- [The method](https://github.com/artefactual/atom/blob/qa/2.x/lib/job/arInformationObjectCsvExportJob.class.php#L133) is called from two places:
  - From [here](https://github.com/artefactual/atom/blob/qa/2.x/apps/qubit/modules/object/actions/exportAction.class.php#L349) without path nor job parameters, which avoids the actual export and returns the count. It should be faster and when the count is "falsey" (no items "exported") it will avoid the actual job execution.
  - From [here](https://github.com/artefactual/atom/blob/qa/2.x/lib/job/arInformationObjectCsvExportJob.class.php#L47) with the parameters, which performs the export and uses the returned count for logging info.

I can't find a good change to improve readability without code duplication, but I'm open to suggestions.

Reply:
Okay, thanks for checking it out @jraddaoui. :)